### DISCUSSION TITLE
How to add field to connection info from `tls_clienthello`

### DISCUSSION THREAD

**Discussion by terryltoms** [mitmproxy/mitmproxy#5888] [Q&A] [answered]
I wrote an addon that implements `tls_clienthello`. The addon calculate a something over the client hello, and I want to show this on the request (connection tab, marker, etc...). How do I do this from the hook?

**Comment by Ilke-dev** [comment]
Hi i think i have a solution for you. The ```ClientHelloData``` class is actually also accessible from the ```HTTPFlow``` class as a client connection.

Checkout this attribute of ```ClientHelloData```: https://github.com/mitmproxy/mitmproxy/blob/main/mitmproxy/tls.py#L120

Then checkout this attribute of ```Flow```: https://github.com/mitmproxy/mitmproxy/blob/main/mitmproxy/flow.py#L56

I think you can access the ClientHelloData from the ```Flow``` class as well.

(FYI ```Flow``` class is inherited by ```HTTPFlow``` that is passed along the hook: https://github.com/mitmproxy/mitmproxy/blob/main/mitmproxy/http.py#L1230)

**Reply by terryltoms** [reply]
I think it wouldn't work.
`clientHelloData -> context -> client`
`flow -> client`
Both hold client, but we cannot get to client hello data from flow. I tried to save a field there, but it is very hacky and didn't work

**Comment by mhils** [comment, marked as answer]
The current workaround to do this is to define a `client_conn -> x` `WeakKeyDictionary`. You can see this in action in the [`ProxyAuth`](https://github.com/mitmproxy/mitmproxy/blob/0b89aede3817831394e6473d82eb04ce572ef621/mitmproxy/addons/proxyauth.py#L28-L30) addon. We should maybe add something like `client_conn.metadata` at some point, let's see. :)
